### PR TITLE
Update Docker setup

### PR DESCRIPTION
**Pull request by MatthewL246**
### Changes:

This updates the Dockerfile to use a new Docker Go template with a few minor changes, including running as a non-root user. This also adds a GitHub Action for building and pushing the Docker image to Docker Hub. This image is currently available at https://hub.docker.com/r/matthewl246/pretendo-friends and will be deleted when this PR is merged.

- [x] I have read and agreed to the [Code of Conduct](https://github.com/PretendoNetwork/Pretendo/blob/master/.github/CODE_OF_CONDUCT.md).
- [x] I have read and complied with the [contributing guidelines](https://github.com/PretendoNetwork/Pretendo/blob/master/.github/CONTRIBUTING.md).
- [x] ~~What I'm implementing was an [approved issue](../issues?q=is%3Aopen+is%3Aissue+label%3Aapproved).~~ This issue was approved internally.
- [x] I have tested all of my changes.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,19 +1,36 @@
...
-COPY . ./
-ARG BUILD_STRING=pretendo.friends.docker
-RUN go build -ldflags "-X 'main.serverBuildString=${BUILD_STRING}'" -v -o server
 
-# --- runner ---
-FROM alpine:3.17 as runner
-WORKDIR /build
+# * Building the application
+FROM golang:1.22-alpine3.20 AS build
+ARG app_dir build_string=pretendo.friends.docker
```
Comment (binaryoverload) [inline Dockerfile]: I personally think it'd be better for clarity to separate these out into two `ARG` calls - they don't cause a new layer to be made so there's no real optimisation here afaik

File: Dockerfile [hunk 2]
```diff
@@ -1,19 +1,36 @@
...
+
+# * Running the final application
+FROM alpine:3.20 AS final
+ARG app_dir
+WORKDIR ${app_dir}
+
+RUN addgroup go && adduser -D -G go go
+
+RUN mkdir -p ${app_dir}/log
+RUN chown go:go ${app_dir}/log
```
Comment (binaryoverload) [inline Dockerfile]: Personally I'd put these in the same run call - won't really matter with an image this size but a small optimisation nonetheless

File: Dockerfile [hunk 3]
```diff
@@ -1,19 +1,36 @@
...
+WORKDIR ${app_dir}
+
+RUN --mount=type=cache,target=/go/pkg/mod/ \
+	--mount=type=bind,source=go.sum,target=go.sum \
+	--mount=type=bind,source=go.mod,target=go.mod \
+	go mod download -x
+
+COPY . .
+RUN --mount=type=cache,target=/go/pkg/mod/ \
+	CGO_ENABLED=0 go build -v -o ${app_dir}/build/server -ldflags "-X 'main.serverBuildString=${build_string}'"
```
Comment (binaryoverload) [inline Dockerfile:20]: Out of curiosity, what is the build string? Is this a Go/Pretendo thing?

**Review by binaryoverload — CHANGES_REQUESTED** [review]

File: Dockerfile (hunk 3 above)
Comment (MatthewL246) [reply Dockerfile:19]: This is a Pretendo-specific thing. @jonbarrow can explain more accurately about what it does.

Comment (MatthewL246) [reply Dockerfile:20]: This is a Pretendo-specific thing. @jonbarrow can explain more accurately about what it does.

**Review by binaryoverload — APPROVED** [review]

File: Dockerfile (hunk 3 above)
Comment (jonbarrow) [reply Dockerfile:20]: @binaryoverload Every game server has 2 servers, the authentication server and the real server (called the "secure server"). The client must first pass through the authentication server before being given the address of the secure server

In the login methods the client uses for the authentication server, the servers build string is sent. It contains information about the server like it's build number, branch, NEX version, etc.

Most games don't care what is set here, but some do scan the build string for relevant data. So it must be set. For example Mario Kart 8 has the following build string `branch:origin/project/nfs build:3_10_26_2006_0`

**Pull request merged**